graphql_compiler/schema_transformation/split_query.py

```diff
@@ -34,6 +39,340 @@ def __init__(self, query_ast):
+    # NOTE: if child_type_name does not actually exist as a root field (not all types are
+    # required to have a corresponding root vertex field), then this query will be invalid
+    child_query_ast = _get_query_document(child_type_name, child_selections)
+    child_query_node = SubQueryNode(child_query_ast)
+
+    return child_query_node, child_output_name
+
+
+def _get_property_field(selections, field_name, directives_from_edge):
+    """Return a Field object with field_name, sharing directives with any such existing field.
+
+    Any valid directives in directives_on_edge will be transferred over to the new field.
+    If there is an existing Field in selection with field_name, the returned new Field
+    will also contain all directives of the existing field with that name.
+
+    Args:
+        selections: List[Union[Field, InlineFragment]]. If there is a field with field_name,
+                    the directives of this field will carry over to the output field. It is
+                    not modified by this function
+        field_name: str, the name of the output field
+        directives_from_edge: List[Directive], the directives of a vertex field. The output
+                              field will contain all @filter and any @optional directives
+                              from this list
+
+    Returns:
+        Field object, with field_name as its name, containing directives from any field in the
+        input selections with the same name and directives from the input list of directives
+    """
+    new_field = Field(
+        name=Name(value=field_name),
+        directives=[],
+    )
+
+    # Check parent_selection for existing field of given name
+    parent_field = try_get_ast_by_name_and_type(selections, field_name, Field)
+    if parent_field is not None:
+        # Existing field, add all its directives
+        directives_from_existing_field = parent_field.directives
+        if directives_from_existing_field is not None:
+            new_field.directives.extend(directives_from_existing_field)
+
+    # Transfer directives from edge
+    if directives_from_edge is not None:
+        for directive in directives_from_edge:
+            if directive.name.value == OutputDirective.name:  # output illegal on vertex field
+                raise GraphQLValidationError(
+                    u'Directive "{}" is not allowed on a vertex field, as @output directives '
+                    u'can only exist on property fields.'.format(directive)
+                )
+            elif directive.name.value == OptionalDirective.name:
+                if try_get_ast_by_name_and_type(
+                    new_field.directives, OptionalDirective.name, Directive
+                ) is None:
+                    # New optional directive
+                    new_field.directives.append(directive)
+            elif directive.name.value == FilterDirective.name:
+                new_field.directives.append(directive)
+            else:
```

Comment:
Could `@fold` or `@output_source` make it to here?

Reply:
Currently, the validation step only allows output, filter, and optional to be in the AST.